ADALiOS/ADALiOS/ADTokenCache.m

```diff
@@ -0,0 +1,546 @@
+    BOOL result = [self removeImpl:item error:error];
+    [_delegate didWriteCache:self];
+    return result;
+}
+
+- (BOOL)removeImpl:(ADTokenCacheItem *)item
+             error:(ADAuthenticationError * __autoreleasing *)error
+{
+    ADTokenCacheKey* key = [item extractKey:error];
+    if (!key)
+    {
+        return NO;
+    }
+    
+    NSString* userId = item.userInformation.userId;
+    if (!userId)
+    {
+        userId = @"";
+    }
+    
+    NSMutableDictionary* tokens = [_cache objectForKey:@"tokens"];
+    if (!tokens)
+    {
+        return YES;
+    }
+    
+    NSMutableDictionary* userTokens = [tokens objectForKey:userId];
+    if (!userTokens)
+    {
+        return YES;
+    }
+    
+    if (![userTokens objectForKey:key])
+    {
+        return YES;
+    }
+    
+    [userTokens removeObjectForKey:key];
+    
+    // Check to see if we need to remove the idtoken
+    if (!userTokens.count)
+    {
+        [tokens removeObjectForKey:userId];
+        [[_cache objectForKey:@"idtokens"] removeObjectForKey:userId];
+    }
+    
+    return YES;
+}
+
+/*! Return a copy of all items. The array will contain ADTokenCacheItem objects,
+ containing all of the cached information. Returns an empty array, if no items are found.
+ Returns nil in case of error. */
+- (NSArray<ADTokenCacheItem *> *)allItems:(ADAuthenticationError * __autoreleasing *)error
+{
+    return [self getItemsWithKey:nil userId:nil error:error];
+}
+
+@end
+
+
+@implementation ADTokenCache (Internal)
+
+- (BOOL)validateCache:(NSDictionary*)dict
+                error:(ADAuthenticationError * __autoreleasing *)error
+{
+    CHECK_ERROR([dict isKindOfClass:[NSDictionary class]], AD_ERROR_BAD_CACHE_FORMAT, @"Root level object of cache is not an NSDictionary!");
+    
+    NSString* version = [dict objectForKey:@"version"];
+    CHECK_ERROR(version, AD_ERROR_BAD_CACHE_FORMAT, @"Missing version number from cache.");
+    CHECK_ERROR([version floatValue] <= CURRENT_WRAPPER_CACHE_VERSION, AD_ERROR_CACHE_PERSISTENCE, @"Cache is a future unsupported version.");
+    
+    NSDictionary* cache = [dict objectForKey:@"tokenCache"];
+    CHECK_ERROR(cache, AD_ERROR_BAD_CACHE_FORMAT, @"Missing token cache from data.");
+    CHECK_ERROR([cache isKindOfClass:[NSMutableDictionary class]], AD_ERROR_BAD_CACHE_FORMAT, @"Cache is not a dictionary!");
+    
+    NSDictionary* tokens = [cache objectForKey:@"tokens"];
+    
+    if (tokens)
+    {
+        CHECK_ERROR([tokens isKindOfClass:[NSMutableDictionary class]], AD_ERROR_BAD_CACHE_FORMAT, @"tokens must be a mutable dictionary.");
+        for (id userId in tokens)
+        {
+            // On the second level we're expecting NSDictionaries keyed off of the user ids (an NSString*) or
+            // NSNull if no userId was available.
+            CHECK_ERROR([userId isKindOfClass:[NSString class]] || [userId isKindOfClass:[NSNull class]], AD_ERROR_BAD_CACHE_FORMAT, @"User ID key not the expected class type");
```

Comment:
I was using NSNull for the user id key when I didn't get a user ID as I can't add things to a dictionary for a nil key. I ended up switching to an empty string (@"") to match the behavior of the keychain cache. I can remove this NSNull check here though.